README.md
# skirmishbunker
python Library for making a skirmish wargame bunker

![](./documentation/image/cover.png)

## Documentation
[Thorough writeup of the development of the bunker](./documentation/README.md)

### Example Usage

``` python
import cadquery as cq
from skirmishbunker import Bunker

bp = Bunker()
bp.inset=15
bp.width=140
bp.length=110
bp.height=65

bp.render_windows=True
bp.skip_windows = []
bp.window_length = 18
bp.window_height = 8
bp.window_frame_chamfer = 1.6
bp.window_frame_chamfer_select = "<Z"

bp.render_doors=True
bp.door_panels = [0, 3]

bp.render_ladders=True
bp.ladder_panels = [8]

bp.render_floor_tiles=True
bp.render_roof=False

bp.make()
rec = bp.build()

cq.exporters.export(rec,'bunker.stl')
```

## Dependencies
* [CadQuery 2.1](https://github.com/CadQuery/cadquery)
* [cqMore](https://github.com/JustinSDK/cqMore)
* [cadqueryhelper](https://github.com/medicationforall/cadqueryhelper)
* [cqterrain](https://github.com/medicationforall/cqterrain)


### Installation
To install skirmishbunker directly from GitHub, run the following `pip` command:

	pip install git+https://github.com/medicationforall/skirmishbunker

**OR**

### Local Installation
From the cloned skirmishbunker directory run.

	pip install ./

### Install a development version

If you want to work on this library locally you will need the following instaled globally on you computer:

* Python 3
* setuptools
* pip
* virtualenv (venv)

Next you will need to clone this repo (or, ideally, your own fork):

	git clone https://github.com/medicationforall/skirmishbunker

Now, step into the cloned folder and build a Python virtual environment:

	cd /path/to/skirmishbunker
	python -m venv ./venv

Once this is done, activate the virtual environment and install dependencies:

	source ./venv/bin/activate
	pip install ./

This will take a little while to run, once finished, you might need to link `skirmishbunker` as an editable library (or your changes will not be seen). Do this as follows:

	pip install -e ./

Now edit away! 

changes.md
## Main wip

## 2.1.0
* Upped cqterrain version to 0.3.0
* Removed local Base class and instead resolve Base from cadqueryhelper

## 2.0.1
* breaking stuff in the DetailedRoof
  * re-wrote how wall cuts care calculated
  * fixed positioning of wall details.
  * Code is still full of nonense and magic constants
* Cleaned up dependencies

## 2.0.0
### Contributed By Morven Lewis-Everley
* Updated flat roof to use hatches and tiles.
* Migrate Roof to Detailed roof, make it inherit from FlatRoof and refactor how tiles, hatches and magnet holes are generated.
* Introduce a helper class that assist in generation of a Series of shapes (as all series code seems to be 90% duplication with only a few changes).
* Added basic dependency injection for swapping out roof type on the fly.
* Added the ability to customise the hatches when generating the roof and the bunker together.
* Added an example of swapping out the roof.

### Cleanup
* Added Apache 2 license to top of source files.
* Restored cut hatch code.
* Restored tile generation code for detailed roof.
  * this was due to a regression bug.
  * temp hack for now
* Fix parameter order when making pip / magnet holes in the roof.
* Make Calc methods protected
* Brought back the render_hatches flag
* Added the render_hatch_cuts flag
* FlatRoof changed hole_diameter to hole_radius
* Added Bunker 5 example.
* Added Bunker 6 example.


## 1.2.0
* New parameter floor_thickness
  * Allows better granular control of the inner dimensions of the bunker.

## 1.1.4
* Updated The License
* Added window width as parameter which can override inset.

## 1.1.3
* Added Split Door
* Increased magnet height from 1.7 to 2.1
* Added catwalk generator.
* Fixed roof magnet cuts.
* Attempt to fix bunker floor placement.
* Updated dependencies
  * cadqueryhelper went to 0.1.1
  * cqterrain went to 0.1.5

## 1.1.2
* Jamie Broke the __init__.py at the root directory of cqterrain

## 1.1.1 Updated version dependencies, added better support for magnets
* Increased Pip Cuts height by .2 mm to try account for magnet sizing
  * Affected Bunker and roof code
* Updated medium bunker to use regular flooring
* Updated Cadquery version
* cadqueryhelper 0.0.8
* cqterrain 0.1.1

## 1.1.0 Customization
* Added support for custom window cuts
* Added support for custom windows
* Custom floor tiles
* Added ladder support
  * Custom Ladders
* Added hatch detailing and hatch cut
* Added floor cuts
* Fix Floor cut with tiles

## 1.0.0 Initial Release
* Date: 2022-12-14
* Initial release of the project


example/bunker_med.py
import cadquery as cq
from skirmishbunker import Bunker
from cqterrain import tile

def custom_star_tile(bunker):
    star_tile = tile.star(
        length = bunker.floor_tile_size,
        width = bunker.floor_tile_size,
        height = bunker.floor_tile_height,
        points = 4,
        outer_radius = bunker.floor_tile_size/2,
        inner_radius = 2,
        padding = .5
    )
    return star_tile

def custom_windmill_tile(bunker):
    windmill_tile = tile.windmill(
        tile_size = bunker.floor_tile_size,
        height = bunker.floor_tile_height,
        padding = .5
    )
    return windmill_tile

bp = Bunker()
bp.inset=15
bp.width=140
bp.length=110
bp.height=65

bp.render_panel_details=True
bp.panel_length=28
bp.panel_width = 5
bp.panel_padding = 4

bp.render_windows=True
bp.skip_windows = []
bp.window_length = 18
bp.window_height = 8
bp.window_frame_chamfer = 1.6
bp.window_frame_chamfer_select = "<Z"

bp.render_doors=True
bp.door_panels = [0]

bp.render_ladders=True
bp.ladder_panels = [8]

bp.render_floor_tiles=True
bp.render_roof=True

bp.floor_padding = -5
bp.floor_tile_padding=.5

bp.render_floor_cuts = False
bp.render_pips=True
bp.render_magnets=True

#bp.custom_floor_tile = custom_windmill_tile

bp.make()

body_complete = bp.build()
cq.exporters.export(body_complete,'stl/bunker_med_both.stl')

body = bp.build_body()
cq.exporters.export(body,'stl/bunker_med_body.stl')

roof = bp.build_roof()
cq.exporters.export(roof,'stl/bunker_med_roof.stl')
